File: wallendar-app/app/src/main/java/com/td/wallendar/group/ui/GroupPresenter.java
package com.td.wallendar.group.ui;

import com.td.wallendar.models.Group;
import com.td.wallendar.models.GroupHistory;
import com.td.wallendar.repositories.interfaces.GroupsRepository;
import com.td.wallendar.utils.scheduler.SchedulerProvider;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.Set;
import java.util.TreeSet;

import io.reactivex.disposables.CompositeDisposable;

public class GroupPresenter {

    private static final Comparator<GroupHistory> groupHistoryComparator = (gh1, gh2) -> gh2.getDate().compareTo(gh1.getDate());

    private final WeakReference<GroupView> groupView;
    private final GroupsRepository groupsRepository;
    private final CompositeDisposable disposable;
    private final SchedulerProvider schedulerProvider;

    public GroupPresenter(final GroupView groupView,
                          final GroupsRepository groupsRepository,
                          final SchedulerProvider schedulerProvider) {
        this.groupView = new WeakReference<>(groupView);
        this.groupsRepository = groupsRepository;
        this.schedulerProvider = schedulerProvider;
        this.disposable = new CompositeDisposable();
    }

    public void onViewAttached(final Long groupId) {
        if (groupsRepository != null) {
            disposable.add(groupsRepository.getGroup(groupId)
                    .subscribeOn(schedulerProvider.io())
                    .observeOn(schedulerProvider.ui())
                    .subscribe(this::onGroupReceived, this::onGroupError));
        }
    }

    private void onGroupReceived(Group group) {
        groupView.get().bindGroup(group);

        Set<GroupHistory> historic = new TreeSet<>(groupHistoryComparator);
        historic.addAll(group.getCharges());
        historic.addAll(group.getPayments());

        groupView.get().bindGroupHistory(new ArrayList<>(historic));
    }

    private void onGroupError(Throwable throwable) {
        System.out.println(throwable);
    }

    public void onViewDetached() {
        disposable.dispose();
    }
}

File: wallendar-app/app/src/main/java/com/td/wallendar/group/ui/GroupPresenter.kt
package com.td.wallendar.group.ui

import androidx.annotation.VisibleForTesting
import com.td.wallendar.models.Group
import com.td.wallendar.models.GroupHistory
import com.td.wallendar.repositories.interfaces.GroupsRepository
import com.td.wallendar.utils.scheduler.SchedulerProvider
import io.reactivex.disposables.CompositeDisposable
import java.lang.ref.WeakReference
import java.util.*

class GroupPresenter(
        groupView: GroupView,
        private val groupsRepository: GroupsRepository,
        private val schedulerProvider: SchedulerProvider
) {
    private val groupView: WeakReference<GroupView> = WeakReference(groupView)
    private val disposable: CompositeDisposable = CompositeDisposable()
    fun onViewAttached(groupId: Long) {
        disposable.add(groupsRepository.getGroup(groupId)
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe({ group: Group -> onGroupReceived(group) }) { throwable: Throwable -> onGroupError(throwable) })
    }

    private fun onGroupReceived(group: Group) {
        groupView.get()?.bindGroup(group)
        val historic: MutableSet<GroupHistory> = TreeSet(groupHistoryComparator)
        historic.addAll(group.charges)
        historic.addAll(group.payments)
        groupView.get()?.bindGroupHistory(ArrayList(historic))
    }

    private fun onGroupError(throwable: Throwable) {
        groupView.get()?.onGroupLoadError()
    }

    fun onViewDetached() {
        disposable.clear()
    }

    fun onViewDestroyed() {
        disposable.dispose()
    }

    @VisibleForTesting
    fun getDisposable(): CompositeDisposable {
        return disposable
    }

    companion object {
        private val groupHistoryComparator: Comparator<GroupHistory> = Comparator { gh1: GroupHistory, gh2: GroupHistory -> gh2.getDate().compareTo(gh1.getDate()) }
    }

}